### PR TITLE
Deprecate support for chained multi-fields.

### PR DESCRIPTION
**Pull request by jtibshirani**
This PR contains a straight backport of #41926, and also updates the
migration documentation and deprecation info API for 7.x.

The code currently contains references to 7.2, but I will update these to 7.3
once we cut the branch over.

### CODE REVIEW

File: x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/IndexDeprecationChecks.java [hunk 1]
```diff
@@ -115,6 +115,32 @@ static DeprecationIssue tooManyFieldsCheck(IndexMetaData indexMetaData) {
...
     }
 
+    static DeprecationIssue chainedMultiFieldsCheck(IndexMetaData indexMetaData) {
+        List<String> issues = new ArrayList<>();
+        fieldLevelMappingIssue(indexMetaData, ((mappingMetaData, sourceAsMap) -> issues.addAll(
+            findInPropertiesRecursively(mappingMetaData.type(), sourceAsMap, IndexDeprecationChecks::containsChainedMultiFields))));
+        if (issues.size() > 0) {
+            return new DeprecationIssue(DeprecationIssue.Level.CRITICAL,
+                "Multi-fields within multi-fields",
+                "https://www.elastic.co/guide/en/elasticsearch/reference/7.2/breaking-changes-7.2.html" +
```
Comment (gwbrown) [inline x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/IndexDeprecationChecks.java]: Historically, we've usually linked to the next major version breaking changes list (8.0 in this case), but this also has a few issues: Until the 8.0 branch is cut, the only way to link to these docs is by using `master` as the version in the URL, which 404s when 8.0 is actually released.

While having this in the breaking changes list for 7.2 and linking to it there would resolve this problem, I don't think it's how we've typically organized the breaking changes list, and might cause issues keeping up to date if we change the deprecation plan.

Basically there's a bunch of problems with how the breaking changes lists and links to them have work at the moment and there's no good options, all I'm pointing out here is that this is different from what we usually do.

File: x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/IndexDeprecationChecks.java [hunk 2]
```diff
@@ -115,6 +115,32 @@ static DeprecationIssue tooManyFieldsCheck(IndexMetaData indexMetaData) {
         return null;
     }
 
+    static DeprecationIssue chainedMultiFieldsCheck(IndexMetaData indexMetaData) {
+        List<String> issues = new ArrayList<>();
+        fieldLevelMappingIssue(indexMetaData, ((mappingMetaData, sourceAsMap) -> issues.addAll(
+            findInPropertiesRecursively(mappingMetaData.type(), sourceAsMap, IndexDeprecationChecks::containsChainedMultiFields))));
+        if (issues.size() > 0) {
+            return new DeprecationIssue(DeprecationIssue.Level.CRITICAL,
```
Comment (gwbrown) [inline x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/IndexDeprecationChecks.java]: Will 8.0 fail to open an index which has chained multi-fields that was created in 7.x? Usually we support all indices created in ($MAJOR-1), even if they use deprecated features.

If 8.0 will be able to open indices created in 7.x that have chained multi-fields, this should be `Level.WARNING` rather than `Level.CRITICAL`.

File: x-pack/plugin/deprecation/src/test/java/org/elasticsearch/xpack/deprecation/IndexDeprecationChecksTests.java [hunk 3]
```diff
@@ -110,6 +112,43 @@ public void testTooManyFieldsCheck() throws IOException {
         assertEquals(0, withDefaultFieldIssues.size());
     }
 
+    public void testChainedMultiFields() throws IOException {
+        XContentBuilder xContent = XContentFactory.jsonBuilder().startObject()
```
Comment (gwbrown) [inline x-pack/plugin/deprecation/src/test/java/org/elasticsearch/xpack/deprecation/IndexDeprecationChecksTests.java]: I'd like this test case to include at least one field that has a non-chained multi-field to verify that the warning message only contains the fields with a problem.

**Review by AthenaEryma — COMMENTED** [review]
Thank you very much for implementing a deprecation check along with the warnings!

I left a couple comments asking for tweaks, but overall that part looks good. I don't see any obvious problems with the changes to the mapping code/tests, but I'm not very familiar with it.

File: x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/IndexDeprecationChecks.java (hunk 2 above)
Comment (jtibshirani) [reply x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/IndexDeprecationChecks.java]: Thanks for pointing this out, I will make sure to update the 8.0 PR to still allow multi-fields on indices created prior to 8.0. With that change, I'll also be able to lower this to `Level.WARNING`.

File: x-pack/plugin/deprecation/src/test/java/org/elasticsearch/xpack/deprecation/IndexDeprecationChecksTests.java (hunk 3 above)
Comment (jtibshirani) [reply x-pack/plugin/deprecation/src/test/java/org/elasticsearch/xpack/deprecation/IndexDeprecationChecksTests.java]: 👍

File: x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/IndexDeprecationChecks.java (hunk 1 above)
Comment (jtibshirani) [reply x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/IndexDeprecationChecks.java]: Would you suggest I change this to `master` for now, for consistency with other deprecation issues we'll be adding? Then I guess we will update all of these links at once when the 8.0 branch is cut?

Comment (gwbrown) [reply x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/IndexDeprecationChecks.java]: Yes, I think that's what I would recommend for now, that way if we make a change to the docs to make the situation better it'll be easier to do all at once.

**Comment by jtibshirani** [comment]
@gwbrown this is now ready for another look.

**Review by AthenaEryma — APPROVED** [review]
LGTM, thanks @jtibshirani!
